pkg/volume/csi/shared_informer_hook.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+// SharedInformerHook provides a way to dynamically watch objects via a SharedInformer.
+type SharedInformerHook struct {
+	informer           cache.SharedIndexInformer
+	objectNamesToWatch map[string]chan watch.Event
+	mu                 sync.RWMutex
+}
+
+// NewSharedInformerHook returns a new SharedInformerHook.
+func NewSharedInformerHook(informer cache.SharedIndexInformer) *SharedInformerHook {
+	return &SharedInformerHook{
+		informer:           informer,
+		objectNamesToWatch: make(map[string]chan watch.Event),
+	}
+}
+
+// RegisterHandlerFuncs adds event handlers to the VolumeAttachmentInformer field.
+// Keep in mind this operation is not idempotent and should be called once.
+func (s *SharedInformerHook) RegisterHandlerFuncs() {
+	s.informer.AddEventHandler(
+		&cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.addEventHandler,
+			UpdateFunc: s.updateEventHandler,
+			DeleteFunc: s.deleteEventHandler,
+		},
+	)
+}
+
+// CreateWatch creates a watch channel for an object with a specific name.
+func (s *SharedInformerHook) CreateWatch(name string, bufferSize int) <-chan watch.Event {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
```

Comment:
Don't we need check if an entry for this attachID already exists?

Reply:
Good catch, thank you.